### PR TITLE
WIP - PagerDuty integration

### PR DESCRIPTION
**Pull request by ganeshrvel**
Added Pagerduty Change tracking

### CODE REVIEW

File: src/robusta/core/model/env_vars.py [hunk 1]
```diff
@@ -64,7 +64,11 @@
 # Timeout for the ping response, before killing the connection. Must be smaller than the interval
 WEBSOCKET_PING_TIMEOUT = int(os.environ.get("WEBSOCKET_PING_TIMEOUT", 30))
 
-TRACE_INCOMING_REQUESTS = bool(os.environ.get("TRACE_INCOMING_REQUESTS", False))
+TRACE_INCOMING_REQUESTS = bool(os.environ.get("TRACE_INCOMING_REQUESTS", True))
```
Comment (arikalon1) [inline src/robusta/core/model/env_vars.py:67]: We don't want to make this the default behavior

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py [hunk 2]
```diff
@@ -96,17 +196,24 @@ def write_finding(self, finding: Finding, platform_enabled: bool):
...
         headers = {"Content-Type": "application/json"}
-        response = requests.post(self.url, json=body, headers=headers)
+        response = requests.post(self.events_url, json=body, headers=headers)
         if not response.ok:
             logging.error(
                 f"Error sending message to PagerDuty: {response.status_code}, {response.reason}, {response.text}"
             )
 
+    def write_finding(self, finding: Finding, platform_enabled: bool):
+        if finding.aggregation_key == "ConfigurationChange/KubernetesResource/Change":
```
Comment (arikalon1) [inline src/robusta/core/sinks/pagerduty/pagerduty_sink.py:210]: I would put this in a constant, used both here and in the playbook.
So, if someone changes it one day, it doesn't break

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py [hunk 3]
```diff
@@ -120,9 +227,39 @@ def __to_unformatted_text(block: BaseBlock) -> str:
...
+    # fetch the changed values from the block
+    @staticmethod
+    def __block_to_changes(block: BaseBlock, enrichment: Enrichment) -> Optional[Dict[str, Any]]:
+        if isinstance(block, KubernetesDiffBlock):
+            operation = enrichment.annotations.get("operation")
+
+            change_count = 0
+            if operation == K8sOperationType.UPDATE:
+                change_count = block.num_modifications
+            elif operation == K8sOperationType.CREATE:
```
Comment (arikalon1) [inline src/robusta/core/sinks/pagerduty/pagerduty_sink.py:256]: I think we don't need the clause for CREATE and DELETE, right?

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py [hunk 4]
```diff
@@ -48,14 +50,112 @@ def __to_pagerduty_status_type(title: str):
...
+
+    @staticmethod
+    def __send_events_to_pagerduty(self, finding: Finding, platform_enabled: bool):
+        custom_details: dict = {}
+
+        links = []
         if platform_enabled:
-            custom_details["🔎 Investigate"] = finding.get_investigate_uri(self.account_id, self.cluster_name)
+            href = finding.get_investigate_uri(self.account_id, self.cluster_name)
+            links.append({
```
Comment (arikalon1) [inline src/robusta/core/sinks/pagerduty/pagerduty_sink.py:137]: I think the local member here isn't needed 
```
            links.append({
                "text": "🔎 Investigate in Robusta",
                "href": finding.get_investigate_uri(self.account_id, self.cluster_name)
            })

```

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py [hunk 5]
```diff
@@ -48,14 +50,112 @@ def __to_pagerduty_status_type(title: str):
         else:
             return "trigger"
 
-    def write_finding(self, finding: Finding, platform_enabled: bool):
+    @staticmethod
+    def __send_changes_to_pagerduty(self, finding: Finding, platform_enabled: bool):
         custom_details: dict = {}
+        links = []
+        if platform_enabled:
+            href = finding.get_investigate_uri(self.account_id, self.cluster_name)
```
Comment (arikalon1) [inline src/robusta/core/sinks/pagerduty/pagerduty_sink.py:58]: No need for the href local member

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py [hunk 6]
```diff
@@ -48,14 +50,112 @@ def __to_pagerduty_status_type(title: str):
...
+
+                description = finding.description
+                changes_count_text = ""
+
+                if operation == K8sOperationType.UPDATE:
+                    if description:
+                        custom_details["Remarks"] = description
+
+                    change_count = changes["change_count"]
+                    changes_count_text = " ({change_count} {changes})".format(change_count=change_count,
```
Comment (arikalon1) [inline src/robusta/core/sinks/pagerduty/pagerduty_sink.py:104]: I would write this as:
```
f" ({change_count} {"change" if change_count == 1 else "changes"})"
```

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py [hunk 7]
```diff
@@ -48,14 +50,112 @@ def __to_pagerduty_status_type(title: str):
...
+                                                                              changes="change" if change_count == 1 else "changes")
+                elif operation == K8sOperationType.CREATE or operation == K8sOperationType.DELETE:
```
Comment (arikalon1) [inline src/robusta/core/sinks/pagerduty/pagerduty_sink.py:106]: I would change this:
```
elif operation == K8sOperationType.CREATE or operation == K8sOperationType.DELETE:
                    if description:
```

to this:
```
elif description:  # operation is CREATE or DELETE
```

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py [hunk 8]
```diff
@@ -48,14 +50,112 @@ def __to_pagerduty_status_type(title: str):
...
+        custom_details["resource"] = f"{finding.service.resource_type}/{finding.subject.name}"
+
+        if finding.subject.node:
+            custom_details["node"] = finding.subject.node
+
+        timestamp = finding.starts_at.astimezone().isoformat()
+
+        for enrichment in finding.enrichments:
+            for block in enrichment.blocks:
+                changes = self.__block_to_changes(block, enrichment)
```
Comment (arikalon1) [inline src/robusta/core/sinks/pagerduty/pagerduty_sink.py:81]: I think it's better to add here:
```
if not isinstance(block, KubernetesDiffBlock):
  continue
```

Then, `__block_to_changes`:
```
    @staticmethod
    def __block_to_changes(block: KubernetesDiffBlock, enrichment: Enrichment) -> Optional[Dict[str, Any]]:
        operation = enrichment.annotations.get("operation")

        change_count = 0
        if operation == K8sOperationType.UPDATE:
            change_count = block.num_modifications

        return {
            "change_count": change_count,
            "operation": operation,
        }
```

Then, no need for:
```
                if not changes:
                    continue
```
only:
```
                operation = changes["operation"]
                if not operation:  # KubernetesDiffBlock diff block without "operation" annotation
                    continue
```

WDYT?

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py [hunk 9]
```diff
@@ -48,14 +50,112 @@ def __to_pagerduty_status_type(title: str):
...
+                if unformatted_texts:
+                    change_num = 1
+                    for diff_text in unformatted_texts:
+                        custom_details[f"Change {change_num}"] = diff_text
+                        change_num += 1
+
+                description = finding.description
+                changes_count_text = ""
+
+                if operation == K8sOperationType.UPDATE:
```
Comment (arikalon1) [inline src/robusta/core/sinks/pagerduty/pagerduty_sink.py:99]: `if operation == K8sOperationType.UPDATE and description:`

**Review by arikalon1 — APPROVED** [review]
Very nice!

Left some minor comments

File: src/robusta/core/model/env_vars.py (hunk 1 above)
Comment (ganeshrvel) [reply src/robusta/core/model/env_vars.py]: done

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py (hunk 2 above)
Comment (ganeshrvel) [reply src/robusta/core/sinks/pagerduty/pagerduty_sink.py]: done

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py (hunk 3 above)
Comment (ganeshrvel) [reply src/robusta/core/sinks/pagerduty/pagerduty_sink.py]: Yes, not required, I have removed them.

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py (hunk 4 above)
Comment (ganeshrvel) [reply src/robusta/core/sinks/pagerduty/pagerduty_sink.py:136]: Cleaned the code up

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py (hunk 5 above)
Comment (ganeshrvel) [reply src/robusta/core/sinks/pagerduty/pagerduty_sink.py]: Cleaned the code up

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py (hunk 9 above)
Comment (ganeshrvel) [reply src/robusta/core/sinks/pagerduty/pagerduty_sink.py]: Done

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py (hunk 8 above)
Comment (ganeshrvel) [reply src/robusta/core/sinks/pagerduty/pagerduty_sink.py]: I wanted made it future proof and generic. I have made the changes as been suggested. Anyway in the future we could make the changes as required.

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py (hunk 7 above)
Comment (ganeshrvel) [reply src/robusta/core/sinks/pagerduty/pagerduty_sink.py]: done

File: src/robusta/core/sinks/pagerduty/pagerduty_sink.py (hunk 6 above)
Comment (ganeshrvel) [reply src/robusta/core/sinks/pagerduty/pagerduty_sink.py]: done

**Pull request merged**